agent/gs_uploader.go

```diff
@@ -90,10 +90,14 @@ func (u *GSUploader) URL(artifact *api.Artifact) string {
 	host := "storage.googleapis.com"
 	path := fmt.Sprintf("%s/%s", u.BucketName, u.artifactPath(artifact))
 
-	// When overriding the default GCS host to go through a proxy, we should use the path passed in directly - it's not safe to assume that a proxy's URL path follows the same path format as GCS.
-	if os.Getenv("BUILDKITE_GCS_ACCESS_HOST") != "" {
-		host = os.Getenv("BUILDKITE_GCS_ACCESS_HOST")
-		path = u.artifactPath(artifact)
+	// Override the default host if required.
+	if envHost, ok := os.LookupEnv("BUILDKITE_GCS_ACCESS_HOST"); ok {
+		host = envHost
+	}
+
+	// If this is set, we trust the user to supply a valid path prefix, and don't use a default GCS path
+	if prefix, ok := os.LookupEnv("BUILDKITE_GCS_PATH_PREFIX"); ok {
```

Comment:
Validate that it starts with a /?

Reply:
It shouldn't start with a `/` - https://golang.org/pkg/net/url/#URL specifies that `/` prefixed paths are treated differently than those without, but that both are valid.  The existing behavior does not have a leading `/`, so I'd rather not change that.

This does, however, raise the valid point that there will always need to be exactly one `/` between `prefix` and `u.artifactPath(artifact)`, so I'll change this to ensure that is always true.